staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go

```diff
@@ -6713,6 +6713,36 @@ func TestUnknownField(t *testing.T) {
 			ExpectedThreeWay:       `{}`,
 			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
 		},
+		"no diff even if null extra": {
+			Original: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Current:  `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":null},"name":"foo","scalar":true}`,
+
+			ExpectedTwoWay:         `{"complex_nullable":{"key":null}}`,
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			ExpectedThreeWay:       `{"complex_nullable":{"key":null}}`,
+			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+		},
+		"discard nulls in nested adds not nulls": {
+			Original: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Current:  `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":{"keynotnull":"value","keynull":null}},"name":"foo","scalar":true}`,
+
+			ExpectedTwoWay:         `{"complex_nullable":{"key":{"keynotnull":"value","keynull":null}}}`,
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":{"keynotnull":"value"}},"name":"foo","scalar":true}`,
+			ExpectedThreeWay:       `{"complex_nullable":{"key":{"keynotnull":"value","keynull":null}}}`,
+			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":{"keynotnull":"value"}},"name":"foo","scalar":true}`,
+		},
+		"discard empty slices": {
+			Original: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Current:  `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":[],"name":"foo","scalar":true}`,
+
+			ExpectedTwoWay:         `{"complex_nullable":[]}`,
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			ExpectedThreeWay:       `{"complex_nullable":[]}`,
+			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+		},
```

Comment:
Do we have a test that shows that we keep null values in the object?

i.e.:
```
Current: `{"something": null}`
ExpectedTwoWay: `{}`
ExpectedTwoWayResult: `{"something": null}
```

Reply:
I tried this preserve test in master branch;

```
"added only": {
			Original: `{"name":"foo"}`,
			Current:  `{"name":"foo","exist":null}`,
			Modified: `{"name":"foo","scalar":true,"complex":{"nested":true},"array":[1,2,3]}`,

			ExpectedTwoWay:         `{"array":[1,2,3],"complex":{"nested":true},"scalar":true}`,
			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}`,
			ExpectedThreeWay:       `{"array":[1,2,3],"complex":{"nested":true},"scalar":true}`,
			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}`,
		},
```

It failed with;

```
using strategicpatch.PatchMetaFromStruct in testcase added only and preserve nulls: expected two-way result:
        	{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}
        got
        	{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}
```

AFAICS, strategic patch removes all null fields during patch from current. This PR aligns this behavior applying also to modified.

Isn't this the correct behavior?, if not maybe this patch should focus on preserving null fields to preserve idempotency?

Reply:
No, the object should be able to carry null values without having these removed by the patch. Null values *in the patch* instruct that the fields should be removed.

If you have an object with `{"a": null, "b": "some value"}`, why would a patch that changes `b` (e.g. `{"b": "new value"}` also remove `a`?

Reply:
@apelisse I definitely agree with you about in the example you gave ^ `a` should not be removed.

However, I checked jsonmergepatch as reference;

```
    original:
      simpleMap:
        key1: 1
        key2: 1
    modified:
      simpleMap:
        key1: 1
    current:
      simpleMap:
        key1: 1
        key2: 1
        other: a
    threeWay:
      simpleMap:
        key2: null
    result:
      simpleMap:
        key1: 1
        other: a
```

`key2` is not in modified but it is removed.

Reply:
Side note/Disclaimer/Rant
---
Yeah, these test confuse me because of the "3 way" vs "2 way" and "modified" vs "current". They don't help test what we're trying to test to be honest.

I guess in most of these tests, I mostly look at "current" "twoway" and "twowayresult" as a way to see how the patch is applied (the 3 way aspects are mostly related to the client AFAIU).

Actual answer :-)
---
So I merely look at "current", "threeway" (the patch), and "result", I see:
The map has 3 fields, the patch asks key2 to be removed, the map has 2 fields with the field remove. Exactly what I would expect. Not sure that's a problem.

Here's something I'd like to see:
```
    original:
      simpleMap:
        key1: 1
    modified:
      simpleMap:
    current:
      simpleMap:
        key1: 1
    threeWay:
      simpleMap:
        key1: null # is that correct?
    result:
      simpleMap:
```

Reply:
Yes, I think your expected result is similar to my findings and I think there is no problem at that part. 

But the problem related to the real issue is in above thread and this PR fixes it. In this thread, mostly we were focusing on "is there any other problem" :).

Reply:
Why I'm saying this because if we filter the nested null values from patchV as this PR suggested. There will no more any problem.